### PR TITLE
EE-1205: add storage::delete to contract API

### PR DESCRIPTION
**Pull request by henrytill**
This PR:
* introduces a `storage::delete` function to the contract API
* introduces similar `RuntimeContext::delete` and `TrackingCopy::delete` methods to support this function

### CODE REVIEW

**Comment by goral09** [comment]
It occurred to me earlier today that in RChain, we were reimbursing users for deleting stuff from the Global State, proportionally to what they paid initially when storing the data. Do we plan on doing the same?

**Comment by goral09** [comment]
> It occurred to me earlier today that in RChain, we were reimbursing users for deleting stuff from the Global State, proportionally to what they had paid initially when storing the data. Do we plan on doing the same?

It's most probably an economics question as well, cc @AlexanderLimonov .

**Review by mpapierski — APPROVED** [review]
Looks good and complete to me (minus one nitpick on tracking copy cache responsibilities)

**Comment by xcthulhu** [comment]
Can a contract delete itself?  I suspect this is a hard business requirement... if so we should have a unit test...

Is it possible for a contract to delete itself and write itself back, and be called twice more?

**Comment by xcthulhu** [comment]
> > It occurred to me earlier today that in RChain, we were reimbursing users for deleting stuff from the Global State, proportionally to what they had paid initially when storing the data. Do we plan on doing the same?
> 
> It's most probably an economics question as well, cc @AlexanderLimonov .



> It occurred to me earlier today that in RChain, we were reimbursing users for deleting stuff from the Global State, proportionally to what they had paid initially when storing the data. Do we plan on doing the same?

Looks like this could be potentially gamed if done naively (see [this article][1] for some more details).

[1]: https://www.coindesk.com/ethereum-fees-gas-token-eip-2751#node-11:~:text=Ethereum%20tries%20to%20mitigate%20this%20problem,gamed%20to%20nab%20lower%20transaction%20fees.

**Comment by goral09** [comment]
> > > It occurred to me earlier today that in RChain, we were reimbursing users for deleting stuff from the Global State, proportionally to what they had paid initially when storing the data. Do we plan on doing the same?
> > 
> > 
> > It's most probably an economics question as well, cc @AlexanderLimonov .
> 
> > It occurred to me earlier today that in RChain, we were reimbursing users for deleting stuff from the Global State, proportionally to what they had paid initially when storing the data. Do we plan on doing the same?
> 
> Looks like this could be potentially gamed if done naively (see [this article](https://www.coindesk.com/ethereum-fees-gas-token-eip-2751#node-11:~:text=Ethereum%20tries%20to%20mitigate%20this%20problem,gamed%20to%20nab%20lower%20transaction%20fees.) for some more details).

Definitely, but if there's no incentive to delete then why would contract/data owner ever do that? Especially, that they have to _pay_ for deletion right now.

File: execution_engine/src/shared/transform.rs [hunk 1]
```diff
@@ -211,6 +212,7 @@ impl Transform {
                 }
             },
             Transform::Failure(error) => Err(error),
+            Transform::Delete => Ok(None),
```
Comment (darthsiroftardis) [inline execution_engine/src/shared/transform.rs:215]: There is a comment in `tracking_copy/mod.rs` saying Transform shouldn't ever produce delete, so i'm confused on what `Transform::Delete` represents here, maybe Im confusing the ordering here???

**Review by darthsiroftardis — APPROVED** [review]
LGTM! There most definitely are nuances that I may have missed, so I'd wait for another greentick!

**Review by xcthulhu — APPROVED** [review]

File: execution_engine/src/core/tracking_copy/mod.rs [hunk 2]
```diff
@@ -255,10 +278,11 @@ impl<R: StateReader<Key, StoredValue>> TrackingCopy<R> {
         correlation_id: CorrelationId,
         key: &Key,
     ) -> Result<Option<StoredValue>, R::Error> {
-        if let Some(value) = self.cache.get(key) {
-            return Ok(Some(value.to_owned()));
-        }
-        if let Some(value) = self.reader.read(correlation_id, key)? {
+        if self.cache.is_deleted(key) {
```
[the remark this answers is not in the document]
Comment (henrytill) [reply execution_engine/src/core/tracking_copy/mod.rs:281]: Perhaps.  Let's opt to address this later when we clean up the `TrackingCopy` and its cache.

**Pull request merged**